cmd/exec_test.go

```diff
@@ -147,3 +148,236 @@ func TestExecBundleFlag(t *testing.T) {
+				"bundle/x.rego": `package system
+
+		test_fun := x {
+			x = false
+			x
+		}
+
+		undefined_test {
+			test_fun
+		}`,
+			},
+			decision:    "",
+			expectError: true,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"error": {
+				"code": "opa_undefined_error",
+				"message": "/system/main decision was undefined"
+			  }
+		}]}`)),
+			failDefined: false,
+			fail:        true,
+		},
+		{
+			description: "--fail with populated result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package system
+
+		main["hello"]`,
+			},
+			decision:    "",
+			expectError: false,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": ["hello"]
+		}]}`)),
+			fail:        true,
+			failDefined: false,
+		},
+		{
+			description: "--fail-defined with true boolean result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package fail.defined.flag
+
+               some_function {
+                       input.foo == 7
+               }
+
+               default fail_test := false
+               fail_test {
+                       some_function
+               }`,
+			},
+			decision:    "fail/defined/flag/fail_test",
+			expectError: false,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": true
+		}]}`)),
+			fail:        true,
+			failDefined: false,
+		},
+		{
+			description: "--fail-defined with false boolean result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package fail.defined.flag
+
+		default fail_test := false
+		fail_test {
+			false
+		}`,
+			},
+			decision:    "fail/defined/flag/fail_test",
+			expectError: false,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": false
+		}]}`)),
+			fail:        true,
+			failDefined: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.description, func(t *testing.T) {
+			test.WithTempFS(tt.files, func(dir string) {
+				var buf bytes.Buffer
+				params := exec.NewParams(&buf)
+				_ = params.OutputFormat.Set("json")
+				params.BundlePaths = []string{dir + "/bundle/"}
+				params.Paths = append(params.Paths, dir+"/files/")
+				if tt.decision != "" {
+					params.Decision = tt.decision
+				}
+				params.FailDefined = tt.failDefined
+				params.Fail = tt.fail
+
+				err := runExec(params)
+				if err != nil && !tt.expectError {
+					t.Fatalf("unexpected error in test: '%s'", tt.description)
+				}
+				if err == nil && tt.expectError {
+					t.Fatalf("expected error, but none occurred in test: '%s'", tt.description)
+				}
+
+				output := util.MustUnmarshalJSON(bytes.ReplaceAll(buf.Bytes(), []byte(dir), nil))
+
+				if !reflect.DeepEqual(output, tt.expected) {
+					t.Fatal("Test", tt.description, "Expected", tt.expected, "Got:", output)
```

Comment:
[nit] We should need the `"Test", tt.description`, since it's nested via `t.Run(tt.description, ...)` already. Also let's do Errorf here:
```suggestion
					t.Errorf("Expected %v, got: %v", tt.expected, output)
```